src/animation.rs

```diff
@@ -0,0 +1,532 @@
+    pub fn action(
+        &mut self,
+        clip: Clip,
+    ) -> Action {
+        let action_data = ActionData::new(clip);
+        let pointer = self.actions.create(action_data);
+        let tx = self.tx.clone();
+        Action { tx, pointer }
+    }
+
+    /// Updates the actions owned by the mixer.
+    pub fn update(
+        &mut self,
+        delta_time: f32,
+    ) {
+        self.process_messages();
+        self.update_actions(delta_time);
+    }
+}
+
+impl ActionData {
+    fn new(clip: Clip) -> Self {
+        ActionData {
+            clip: clip,
+            enabled: true,
+            loop_mode: LoopMode::Repeat { limit: None },
+            paused: false,
+            local_time: 0.0,
+            local_time_scale: 1.0,
+        }
+    }
+
+    /// Updates a single animation action.
+    fn update(
+        &mut self,
+        delta_time: f32,
+    ) {
+        if self.paused || !self.enabled {
+            return;
+        }
+
+        self.local_time += delta_time * self.local_time_scale;
+        let mut finish_count = 0;
+        for mut track in &mut self.clip.tracks {
+            let frame_index = match track.frame_at_time(self.local_time) {
+                FrameRef::Unstarted => continue,
+                FrameRef::Ended => {
+                    finish_count += 1;
+                    continue;
+                }
+                FrameRef::InProgress(i) => i,
+            };
+            let frame_start_time = track.times[frame_index];
+            let frame_end_time = track.times[frame_index + 1];
+            let frame_delta_time = frame_end_time - frame_start_time;
+            // Interpolation constant in range `[0.0, 1.0]` between `frame[i]`
+            // and `frame[i + 1]`.
+            let s = (self.local_time - frame_start_time) / frame_delta_time;
+
+            match (track.binding, &track.values) {
+                (Binding::Orientation, &Values::Euler(ref values)) => {
+                    let frame_start_value = {
+                        let euler = values[frame_index];
+                        cgmath::Quaternion::from(
+                            cgmath::Euler::new(
+                                cgmath::Rad(euler.a),
+                                cgmath::Rad(euler.b),
+                                cgmath::Rad(euler.c),
+                            ),
+                        )
+                    };
+                    let frame_end_value = {
+                        let euler = values[frame_index + 1];
+                        cgmath::Quaternion::from(
+                            cgmath::Euler::new(
+                                cgmath::Rad(euler.a),
+                                cgmath::Rad(euler.b),
+                                cgmath::Rad(euler.c),
+                            ),
+                        )
+                    };
+                    let update = frame_start_value.slerp(frame_end_value, s);
+                    track.object.set_orientation(update);
```

Comment:
We mutate `track.object` here, which is why `&mut self.clip.tracks` is required.